ADEN-7329 - disable FV autoplay based on Queen model prediction

jacekbj opened the pull request.

Commit:
ADEN-7329

dmnsobczak commented on extensions/wikia/AdEngine/js/context/adContext.js:
```diff
@@ -93,6 +93,8 @@ define('ext.wikia.adEngine.adContext', [
 		context.rabbits.ctpDesktop = isProperGeoAds('wgAdDriverCTPDesktopRabbitCountries');
 		context.rabbits.ctpMobile = isProperGeoAds('wgAdDriverCTPMobileRabbitCountries');
+		context.rabbits.queenDesktop = isProperGeoAds('wgAdDriverCTPDesktopQueenCountries');
+		context.rabbits.isQueenDesktopForced = window.location.href.indexOf('CTPQueenForce') > -1;
```
How about this?
`context.rabbits.isQueenDesktopForced = isUrlParamSet('CTPQueenForce');`

jacekbj replied:
beautiful :)

Commit:
ADEN-7329 CR fix

fraszczakszymon commented on extensions/wikia/ArticleVideo/scripts/featured-video.jwplayer.autoplay.js:
```diff
@@ -2,23 +2,28 @@ define('wikia.articleVideo.featuredVideo.autoplay', [
-	}
+		isAutoplayDisabledByRabbits: function () {
+			var isAutoplayDisabledByRabbit = adContext.get('rabbits.ctpDesktop');
+			var queenPrediction = rabbit.getPredictions(['queendesktop']);
```
What do you think about implementing `getPrediction(modelName)` so it returns prediction for given model and you don't need to check length of returned array/use indexOf for checking the result

jacekbj replied:
Good point - I was blindly copying the `getResults` method here.

fraszczakszymon commented on extensions/wikia/ArticleVideo/scripts/featured-video.jwplayer.autoplay.js:
```diff
@@ -2,23 +2,28 @@ define('wikia.articleVideo.featuredVideo.autoplay', [
-		isAutoplayEnabled: isAutoplayEnabled,
-		inNextVideoAutoplayEnabled: inNextVideoAutoplayEnabled
-	}
+		isAutoplayDisabledByRabbits: function () {
```
It seems we don't need to expose this method

jacekbj replied:
We need to if we want to test it :/

fraszczakszymon commented on extensions/wikia/ArticleVideo/scripts/featured-video.jwplayer.autoplay.js:
```diff
@@ -2,23 +2,28 @@ define('wikia.articleVideo.featuredVideo.autoplay', [
 	'wikia.abTest',
 	'wikia.articleVideo.featuredVideo.cookies',
-], function (adContext, abTest, featuredVideoCookieService) {
-	var inFeaturedVideoClickToPlayABTest = abTest.inGroup('FV_CLICK_TO_PLAY', 'CLICK_TO_PLAY');
```
It was fine to check it once for this module

fraszczakszymon commented on extensions/wikia/ArticleVideo/scripts/featured-video.jwplayer.autoplay.js:
```diff
@@ -2,23 +2,28 @@ define('wikia.articleVideo.featuredVideo.autoplay', [
+	'use strict';
 
 	return {
-		isAutoplayEnabled: isAutoplayEnabled,
```
Why? :) Wasn't it cleaner? (having returned interface short)

jacekbj replied:
It was cleaner but also impossible to test. For a method which takes into account X boolean variables I'd have to write X^2 test cases.

Commit:
ADEN-7329 CR fix 2

Commit:
ADEN-7329 CR fix 3 - refactor getPredictions -> getPrediction.

Commit:
ADEN-7329 CR fix 4 - generic URL param to override model prediction.

fraszczakszymon committed:
ADEN-7329 Move rabbits to top scripts

Commit:
ADEN-7329 CR fix 5 - move overriding model output.

fraszczakszymon commented on extensions/wikia/AdEngine/js/context/adContext.js:
```diff
@@ -93,6 +93,7 @@ define('ext.wikia.adEngine.adContext', [
 	function updateAdContextRabbitExperiments(context) {
 		context.rabbits.ctpDesktop = isProperGeoAds('wgAdDriverCTPDesktopRabbitCountries');
 		context.rabbits.ctpMobile = isProperGeoAds('wgAdDriverCTPMobileRabbitCountries');
+		context.rabbits.queenDesktop = isProperGeoAds('wgAdDriverCTPDesktopQueenCountries');
```
This variable has to be added to `AdEngine2Hooks.class.php`

fraszczakszymon commented on extensions/wikia/AdEngine/js/ml/ctp/queenDesktop.js:
```diff
@@ -0,0 +1,25 @@
+			/*
+			Enable only on page with FV on oasis with queen enabled.
+			 */
+			return (
```
Not needed brackets :) (comment too)

fraszczakszymon commented on extensions/wikia/AdEngine/js/ml/ctp/queenDesktopDataSource.js:
```diff
@@ -0,0 +1,8 @@
+/*global define*/
+define('ext.wikia.adEngine.ml.ctp.queenDesktopDataSource', [
```
Can we reverse it? According to AC:

> it can use the same data source for now (ctpDesktopDataSource.js) - rename it to queenOfHearts.js and alias ctpDesktop

Let queen to be the main source

fraszczakszymon commented on extensions/wikia/AdEngine/js/ml/modelFactory.js:
```diff
@@ -48,7 +49,17 @@ define('ext.wikia.adEngine.ml.modelFactory', [
+
 			predict: function () {
+				var key = this.buildModelForcedKey(modelData.name);
+				var qs = new Querystring();
```
`qs` can be created once in `create` method

fraszczakszymon commented on extensions/wikia/AdEngine/js/ml/rabbit.js:
```diff
@@ -33,13 +36,34 @@ define('ext.wikia.adEngine.ml.rabbit', [
+	 */
+	function getPrediction(modelName) {
+		var model = getEnabledModels()
+			// Oh God I miss arrow functions
```
Remove comment please

fraszczakszymon commented on extensions/wikia/AdEngine/js/ml/rabbit.js:
```diff
@@ -33,13 +36,34 @@ define('ext.wikia.adEngine.ml.rabbit', [
+	 * @return {number | undefined} Prediction value or undefined if model is not enabled
+	 */
+	function getPrediction(modelName) {
+		var model = getEnabledModels()
```
s/model/models

fraszczakszymon commented on extensions/wikia/ArticleVideo/scripts/featured-video.jwplayer.autoplay.js:
```diff
@@ -2,23 +2,28 @@ define('wikia.articleVideo.featuredVideo.autoplay', [
-		inNextVideoAutoplayEnabled: inNextVideoAutoplayEnabled
-	}
+		isAutoplayDisabledByRabbits: function () {
+			var queenPrediction = rabbit.getPrediction('queendesktop');
```
What if `rabbits === undefined`? We should check whether `rabbit` module is available here. Or remove `require.optional`

Commit:
ADEN-7329 # CR fix 6.

Commit:
ADEN-7329 # CR fixes 7.

dmnsobczak commented on extensions/wikia/AdEngine/js/ml/modelFactory.js:
```diff
@@ -11,6 +11,7 @@ define('ext.wikia.adEngine.ml.modelFactory', [
 		'name',
 		'enabled'
 	];
+	var qs = new Querystring();
```
If `create()` will be called only once, isn't it better idea to place it there? Correct me if I'm wrong but `qs` object will be created on every pageview no matter if it will or won't execute model, right?

jacekbj replied:
Good point.

fraszczakszymon replied:
+1 it has to be done in `create` because of mobile-wiki

dmnsobczak commented on extensions/wikia/ArticleVideo/scripts/featured-video.jwplayer.autoplay.js:
```diff
@@ -17,10 +17,13 @@ define('wikia.articleVideo.featuredVideo.autoplay', [
 		isAutoplayEnabled: function () {
-			return featuredVideoCookieService.getAutoplay() !== '0' &&
+			var isEnabled =  featuredVideoCookieService.getAutoplay() !== '0' &&
 				// adContext.isEnabled('wgArticleVideoAutoplayCountries') &&
```
Add `ToDo` why it is commented out?

jacekbj replied:
I didn't add a TODO as it was already here, probably a remnant of other team's experiment.

Commit:
ADEN-7398 # QA fix 7.

Enable rabbit also on mobile.

Commit:
ADEN-7398 # QA fix 8.

Disable toggle in FV when rabbit or queen is active.
Remove all checks for mobile in FV - this code is used only on oasis.
